### PR TITLE
Cost function class

### LINKED ISSUE #489: Write a cost function class
Write a class that allows different cost-functions to be used (for example, Chi-Squared (nlls), weighted chi-squared)

### PR DESCRIPTION
**Pull request by wathen**
#### Description of Work

Fixes #489 

Adds new cost function class and removes mention of `r_norm`

#### Testing Instructions

1. Check class does as expected
2. Check there is no mention of `r_norm`

Function: Does the change do what it's supposed to?

Tests: Does it pass? Is there adequate coverage for new code?

Style: Is the coding style consistent? Is anything overly confusing?

Documentation: Is there a suitable change to documentation for this change?

### CODE REVIEW

File: docs/source/extending/cost_function.rst [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+1. Create ``fitbenchmarking/cost_func/<cost_func>_cost_func.py``,
+   which contains a new subclass of
+   :class:`~fitbenchmarking.cost_func.base_cost_func.CostFunc`.
+   Then implement the method:
+
+    -  .. automethod:: fitbenchmarking.cost_func.base_cost_func.CostFunc.eval_cost()
+              :noindex:
+
+2. Document the available cost functions by by:
```
Comment (wathen) [inline docs/source/extending/cost_function.rst:20]:
```suggestion
2. Document the available cost functions by:
```

File: docs/source/extending/jacobian_extend.rst [hunk 2]
```diff
@@ -29,17 +34,23 @@ you will need to:
    ``fitbenchmarking/jacobian/tests/test_jacobians.py``.
 
 
-The :class:`~fitbenchmarking.parsing.fitting_problem.FittingProblem`
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The :class:`~fitbenchmarking.parsing.fitting_problem.FittingProblem` and :class:`~fitbenchmarking.cost_func.base_cost_func.CostFunc`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-When adding new Jacobian, you will find it helpful to make use of the
+When adding new cost functions, you will find it helpful to make use of the
```
Comment (wathen) [inline docs/source/extending/jacobian_extend.rst:40]:
```suggestion
When adding new Jacobian, you will find it helpful to make use of the
```

File: docs/source/extending/jacobian_extend.rst [hunk 3]
```diff
@@ -29,17 +34,23 @@ you will need to:
...
 following members of the :class:`~fitbenchmarking.parsing.fitting_problem.FittingProblem`
-classes:
+and subclasses of :class:`~fitbenchmarking.cost_func.base_cost_func.CostFunc` class:
```
Comment (wathen) [inline docs/source/extending/jacobian_extend.rst:42]:
```suggestion
and subclasses of :class:`~fitbenchmarking.cost_func.base_cost_func.CostFunc`:
```

File: docs/source/users/options/fitting_option.rst [hunk 4]
```diff
@@ -113,3 +113,20 @@ Default is ``scipy``
...
    Currently analytic Jacobians are available are only available for
    problems that use the cutest and NIST parsers.
+
+Cost function (:code:`cost_func_type`)
+--------------------------------------
+
+This sets the cost function to be used for the given data. Current cost
+functions supported are:
+
+* ``nlls`` - This sets the cost function to be weighted or non-weighted non-linear least squares.
```
Comment (wathen) [inline docs/source/users/options/fitting_option.rst:123]:
```suggestion
* ``nlls`` - This sets the cost function to be weighted or non-weighted non-linear least squares, the weighting is controlled by the ``use_errors`` option.
```

File: fitbenchmarking/cost_func/nlls_cost_func.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
+"""
+Implements the base class for the cost function class.
```
Comment (wathen) [inline fitbenchmarking/cost_func/nlls_cost_func.py:2]:
```suggestion
Implements the weighted and non-weighted non-linear least squares cost function```

File: fitbenchmarking/jacobian/base_jacobian.py [hunk 6]
```diff
@@ -14,64 +11,64 @@ class Jacobian:
...
         """
-        if array_equal(params, cached_dict['params']):
-            value = cached_dict['value']
-        else:
-            value = eval_func(params, **kwargs)
-        return value
+        return NotImplementedError
 
-    def eval_r_norm(self, params, **kwargs):
+    def cached_func_values(self, cached_dict, eval_modelunc, params, **kwargs):
```
Comment (wathen) [inline fitbenchmarking/jacobian/base_jacobian.py:53]:
```suggestion
    def cached_func_values(self, cached_dict, eval_model, params, **kwargs):
```

File: fitbenchmarking/jacobian/base_jacobian.py [hunk 7]
```diff
@@ -14,64 +11,64 @@ class Jacobian:
...
-                                     self.problem.eval_r,
-                                     params,
-                                     **kwargs)
-        J = self.eval(params, **kwargs)
-        out = 2.0 * matmul(J.T, rx)
-        return out
+        if array_equal(params, cached_dict['params']):
+            value = cached_dict['value']
+        else:
+            value = eval_modelunc(params, **kwargs)
```
Comment (wathen) [inline fitbenchmarking/jacobian/base_jacobian.py:70]:
```suggestion
            value = eval_model(params, **kwargs)
```

File: fitbenchmarking/parsing/tests/test_parsers.py [hunk 8]
```diff
@@ -246,9 +246,7 @@ def test_function_evaluation(self, file_format, evaluations_file):
 
             for r in tests:
                 x = np.array(r[0])
-                actual = fitting_problem.eval_f(x=x,
-                                                params=r[1])
-                print(r, actual)
+                actual = fitting_problem.function(x, *r[1])
```
Comment (wathen) [inline fitbenchmarking/parsing/tests/test_parsers.py:249]:
```suggestion
                actual = fitting_problem.eval_f(x=x, params=r[1])
```

**Review by tyronerees — CHANGES_REQUESTED** [review]
I don't think the two files `.coverage` and `file` should be in the repository.

I think we need to document this a little more -- it's not clear to me what should go in `eval_r` or in `eval_cost`.  

Is `eval_r` always 
image
If so it should (a) be in the `FittingProblem` class, and (b) that would mean we couldn't solve, say, weighted non-linear least squares using a non-linear least squares solver (unless we put another layer on top, which is possible).

Otherwise, we could say that it solves
image
where eval_cost is c(.) and eval_r is r(.).  This would let us solve problems like 
image
and 
image
using Levenberg-Marquardt type methods (say).  However, there are some methods that can cope with weighting, say, internally (Mantid and RALFit, or the top of my head), and this would add additional overhead.  We should discuss.

I checked this branch against master with the NIST test set and all Mantid (function evaluation only), Sci_ls (eval_r) and minuit (eval_cost), and the accuracy results were identical.

**Review by tyronerees — APPROVED** [review]

**Comment by tyronerees** [comment]
Will update the docs in #490

**Pull request merged**